device_backends/xdma/src/XdmaBackend.cc

```diff
@@ -0,0 +1,126 @@
+#include "XdmaBackend.h"
+
+#include <boost/make_shared.hpp>
+#include <iomanip>
+
+namespace ChimeraTK {
+
+  XdmaBackend::XdmaBackend(std::string devicePath, std::string mapFileName)
+  : NumericAddressedBackend(mapFileName), _devicePath(devicePath) {}
+
+  XdmaBackend::~XdmaBackend() {}
+
+  void XdmaBackend::open() {
+#ifdef _DEBUG
+    std::cout << "open xdma dev: " << _devicePath << std::endl;
+#endif
+    if(_ctrlIntf) {
+      if(isFunctional()) {
+        return;
+      }
+      close();
+    }
+    // TODO: retrieve mmap_(offs,size) from map file
+    _ctrlIntf.emplace(_devicePath);
+
+    // Build vector of max. 4 DMA channels
+    _dmaChannels.clear();
+    for(size_t i = 0; i < 4; i++) {
+      try {
+        _dmaChannels.emplace_back(_devicePath, i);
+      }
+      catch(const runtime_error&) {
+        break;
+      }
+    }
+    _hasActiveException = false;
+  }
+
+  void XdmaBackend::close() {
+    _ctrlIntf.reset();
+    _dmaChannels.clear();
+  }
+
+  bool XdmaBackend::isOpen() { return !!_ctrlIntf; }
+
+  bool XdmaBackend::isFunctional() const {
+    if(!_ctrlIntf || _hasActiveException) {
+      return false;
+    }
+    // TODO implement me
```

Comment:
Seems sufficient to me. The next open will detect this and try a re-connect.

Reply:
Well, yes. You could actively try to read something to see if the device really is still working.